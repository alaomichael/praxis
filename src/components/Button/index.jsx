import React from 'react'
import StyledButton from './styles'
import PropTypes from 'prop-types'
import HeaderFont from './../Typography/Header'

const Button = props => (
    <StyledButton
        {...props}
        disabled={(props.disabled || props.loading)}
        onClick={ e => {
            e.stopPropagation()
            props.onEvent({
                type: 'onClick',
                origin: 'Button',
                data: props.data
            })
        }}
    >
       {
        props.loading && props.loadingLabel ? props.loadingLabel : props.label
       }

    </StyledButton>
)

Button.propTypes = {
    disabled: PropTypes.bool,
    loading: PropTypes.bool,
    label: PropTypes.string.isRequired,
    loadingLabel: PropTypes.string,
    xs: PropTypes.bool,
    sm: PropTypes.bool,
    md: PropTypes.bool,
    lg: PropTypes.bool,
    success: PropTypes.bool,
    primary: PropTypes.bool,
    danger: PropTypes.bool,
    secondary: PropTypes.bool,
    data: PropTypes.object,
    onEvent: PropTypes.func.isRequired
}

Button.defaultProps = {
    disabled: false,
    loading: false,
    loadingLabel: null,
    xs: false,
    sm: false,
    md: false,
    lg: false,
    success: false,
    primary: false,
    danger: false,
    secondary: false,
    data: {}
}

export default Button