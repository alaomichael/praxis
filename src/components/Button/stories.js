
import React from 'react'
import { storiesOf } from '@storybook/react'
import Container, { Row, Column } from '../Grid'
import HeaderText from '../Typography/Header'
import { object, text } from "@storybook/addon-knobs";

import Button from '.'

const label = 'Styles';
const defaultValue = {
  backgroundColor: 'red',
};
const groupId = 'GROUP-ID1';


const stories = storiesOf('Components|Buttons', module)
// stories.addDecorator(withKnobs)
  .add('Default', () => (
    <Container>
        <Row>
            <Column>
                <HeaderText h1 label={text("title", "Buttons")} />
            </Column>
        </Row>
        <Row>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                    data={object(label, defaultValue, groupId)}
                    label={"Default with data"} />
            </Column>
        </Row>
        <Row>
            <Column>
                <HeaderText h2 label='Button sizes' />
            </Column>
        </Row>
        <Row>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                xs label={"Button size xs"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                sm label={"Button size sm"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                md label={"Button size md"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                lg label={"Button size lg"} />
            </Column>
        </Row>
        <Row>
            <Column>
                <HeaderText h2 label='Button Default loading' />
            </Column>
        </Row>
        <Row>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                xs loading label={"Button size xs"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                sm loading label={"Button size sm"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                md loading label={"Button size md"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                lg loading label={"Button size lg"} />
            </Column>
        </Row>
        <Row>
            <Column>
                <HeaderText h2 label='Button Default disabled' />
            </Column>
        </Row>
        <Row>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                xs disabled label={"Button size xs"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                sm disabled label={"Button size sm"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                md disabled label={"Button size md"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                lg disabled label={"Button size lg"} />
            </Column>
        </Row>
        <Row>
            <Column>
                <HeaderText h2 label='Button colorful' />
            </Column>
        </Row>
        <Row>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                xs primary label={"Button primary size xs"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                sm primary label={"Button primary size sm"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                md primary label={"Button primary size md"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                lg primary label={"Button primary size lg"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                lg primary disabled label={"Button primary size lg disabled"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                lg primary loading label={"Button primary size lg loading"} loadingLabel={"Applying..."} />
            </Column>
        </Row>
        <Row>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                xs danger label={"Button danger size xs"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                sm danger label={"Button danger size sm"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                md danger label={"Button danger size md"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                lg danger label={"Button danger size lg"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                lg danger disabled label={"Button danger size lg disabled"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                lg danger loading label={"Button danger size lg loading"}  loadingLabel={"Applying..."} />
            </Column>
        </Row>
        <Row>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                xs success label={"Button success size xs"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                sm success label={"Button success size sm"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                md success label={"Button success size md"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                lg success label={"Button success size lg"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                lg success disabled label={"Button success size lg disabled"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                lg success loading label={"Button success size lg loading"}  loadingLabel={"Applying..."}  />
            </Column>
        </Row>
        <Row>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                xs secondary label={"Button secondary size xs"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                sm secondary label={"Button secondary size sm"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                md secondary label={"Button secondary size md"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                lg secondary label={"Button secondary size lg"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                lg secondary disabled label={"Button secondary size lg disabled"} />
            </Column>
            <Column>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                lg secondary loading label={"Button secondary size lg loading"}  loadingLabel={"Applying..."} />
            </Column>
        </Row>
        <Row>
            <Column>
                <HeaderText h1 label='Button full size' />
            </Column>
        </Row>
        <Row>
            <Column full>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                full label={"Default full size"} />
            </Column>
        </Row>
        <Row>
            <Column full>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                full xs label={"xs full size"} />
            </Column>
        </Row>
        <Row>
            <Column full>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                full sm label={"sm full size"} />
            </Column>
        </Row>
        <Row>
            <Column full>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                full md label={"md full size"} />
            </Column>
        </Row>
        <Row>
            <Column full>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                full lg label={"lg full size"} />
            </Column>
        </Row>
        <Row>
            <Column full>
                <Button
                    onEvent={ e => {
                        console.log(e)
                    }}
                full lg loading label={"lg full size loading"}  loadingLabel={"Applying..."} />
            </Column>
        </Row>
    </Container>
  ))


