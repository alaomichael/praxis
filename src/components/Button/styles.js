import styled, { keyframes } from 'styled-components'
import Colors from '../Colors'

const buttonLoadingAnimation = keyframes`
    to {
        transform: rotate(360deg)
    }
`
const generateButtonColor = color => `
    border: solid 1px;
    border-color: rgba(${Colors[color].r},${Colors[color].g},${Colors[color].b}, 0.8);
    background-color: rgb(${Colors[color].r},${Colors[color].g},${Colors[color].b});
    color: ${Colors[color === "Pink" ?  "White" : "Black"].hex};

    ${
        color === 'White' ? `
        color: rgb(${Colors["Black"].r},${Colors["Black"].g},${Colors["Black"].b});
        border-color: rgba(${Colors["Black"].r},${Colors["Black"].g},${Colors["Black"].b}, 0.8);
        :hover {
            background-color: ${Colors['Light-Blue'].hex};
            border: solid 1px ${Colors['Light-Blue'].hex};
            color: ${Colors["White"].hex};

        }

        :active {
            background-color: rgba(${Colors["Light-Blue"].r},${Colors["Light-Blue"].g},${Colors["Light-Blue"].b}, 0.5);
            border: solid 1px ${Colors[`${color}`].hex};
            color: ${Colors["White"].hex};
        }


        ` : `
        :hover {
            background-color: rgba(${Colors[color].r},${Colors[color].g},${Colors[color].b}, 0.7);
            border: solid 1px ${Colors[`${color}`].hex};
            color: ${Colors["White"].hex};
        }

        :active {
            background-color: rgba(${Colors[color].r},${Colors[color].g},${Colors[color].b}, 0.8);
            border: solid 1px ${Colors[`${color}`].hex};
            color: ${Colors["White"].hex};
        }
        `
    }
`
const generateDisabledButton = props => {
    let color= 'Light-Blue'

    switch ( true ) {
        case props.secondary:
        case props.pink:
            color= 'Pink'
            break
        case props.success:
            color= 'Green'
            break
        case props.white:
            color= 'White'
            break
        case props.danger:
        case props.yellow:
            color= 'Yellow'
            break
        default:
            color= 'Light-Blue'
    }

    return (`

    background: transparent;
    border: 1px solid ;
    border-color: rgba(${Colors[color].r},${Colors[color].g},${Colors[color].b}, 0.4);
    color: rgba(${Colors[color].r},${Colors[color].g},${Colors[color].b}, 0.4);

    :hover {
        background: transparent;
        color: rgba(${Colors[color].r},${Colors[color].g},${Colors[color].b}, 0.4);
        border: 1px solid ;
        border-color: rgba(${Colors[color].r},${Colors[color].g},${Colors[color].b}, 0.4);
        cursor: ${props.loading ? 'wait' : 'not-allowed'};
    }

`)
}

export default styled.a`
    display: inline-block;
    padding: 1.25rem 4rem;
    text-decoration: none;
    text-transform: uppercase;
    justify-content: center;

    white-space: nowrap;

    cursor: pointer;
    position: relative;
    background: transparent;

    text-align: center;


    ${({ xs }) => xs && `
        padding: 1em 3rem;
    `}

    ${({ sm }) => sm && `
        padding: 1em 3rem;
    `}

    ${({ md }) => md && `
        padding: 1em 3rem;
    `}

    ${({ lg }) => lg && `
        padding: 1em 3rem;
    `}

    ${({ full }) => full && `
        width: 100%;
    `}

    :before {
        content: '';
        box-sizing: border-box;
        position: absolute;
        top: 50%;
        left: 50%;
        width: 20px;
        height: 20px;
        margin-top: -10px;
        margin-left: -10000px;
        border-radius: 50%;
        border: 1px solid ${Colors['Pink'].hex};
        border-top-color: ${Colors['Light-Blue'].hex};
        animation: ${buttonLoadingAnimation} .6s linear infinite;
        ${({ loading }) => loading && `
            margin-left: -10px;
        `}
      }

      ${({ primary }) => primary && `
        ${generateButtonColor('Light-Blue')}
      `}

      ${({ blue }) => blue && `
        ${generateButtonColor('Light-Blue')}
      `}

      ${({ danger }) => danger && `
        ${generateButtonColor('Yellow')}
      `}

      ${({ success }) => success && `
        ${generateButtonColor('Green')}
      `}

      ${({ secondary }) => secondary && `
        ${generateButtonColor('Pink')}
      `}

      ${({ pink }) => pink && `
        ${generateButtonColor('Pink')}
      `}

      ${({ yellow }) => yellow && `
        ${generateButtonColor('Yellow')}
      `}

      ${({ white }) => white && `
        ${generateButtonColor('White')}
      `}

      ${props  => props.disabled && `
        ${generateDisabledButton(props)}
      `}


`
