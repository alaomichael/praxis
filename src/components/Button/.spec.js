import React from 'react'
import renderer from 'react-test-renderer'
import Button from '.'

describe('Button', () => {
    it ('Rendering Button without crashing', () => {
        const result = renderer.create(
            <Button
                label={"Test Button"}
                onEvent={e => {
                }}
            />
        ).toJSON()
        expect(result).toMatchSnapshot()
    })
})
