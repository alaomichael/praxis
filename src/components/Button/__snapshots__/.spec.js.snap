// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Rendering Button without crashing 1`] = `
.c0 {
  padding: 12px;
  border-radius: 3px;
  font-size: 14px;
  cursor: pointer;
  position: relative;
}

.c0:before {
  content: '';
  box-sizing: border-box;
  position: absolute;
  top: 50%;
  left: 50%;
  width: 20px;
  height: 20px;
  margin-top: -10px;
  margin-left: -10000px;
  border-radius: 50%;
  border: 1px solid #B0BEC5;
  border-top-color: #2196F3;
  -webkit-animation: hzglfX .6s linear infinite;
  animation: hzglfX .6s linear infinite;
}

<button
  className="c0"
  data={Object {}}
  disabled={false}
  label="Test Button"
  loading={false}
  onClick={[Function]}
  onEvent={[Function]}
>
  Test Button
</button>
`;
